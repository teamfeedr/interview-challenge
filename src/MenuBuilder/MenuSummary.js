import React from 'react';

const MenuSummary = ({ count = 0, dietaries = {} }) => (
  <div className="menu-summary">
    <div className="container">
      <div className="row">
        <div className="col-6 menu-summary-left">
          <span>{count} items</span>
        </div>
        <div className="col-6 menu-summary-right">
          {Object.keys(dietaries).map(
            diet =>
              !!dietaries[diet] && (
                <span key={diet}>
                  {dietaries[diet]}x <span className="dietary">{diet}</span>
                </span>
              )
          )}
        </div>
      </div>
    </div>
  </div>
);

export default MenuSummary;
