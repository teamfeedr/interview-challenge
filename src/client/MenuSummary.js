import React from 'react';

export default ({ selectedItems }) => {
  const dietaryMap = selectedItems.reduce((acc, { dietaries }) => {
    dietaries.forEach((label) => {
      if (acc[label]) {
        acc[label] += 1;
      } else {
        acc[label] = 1;
      }
    });

    return acc;
  }, {});

  return (
    <div className="menu-summary">
      <div className="container">
        <div className="row">
          <div className="col-6 menu-summary-left">
            <span>{`${selectedItems.length} item${selectedItems.length === 1 ? '' : 's'}`}</span>
          </div>
          <div className="col-6 menu-summary-right">
            {Object.entries(dietaryMap).map(([label, count]) => (
              <span key={label}>{count}x <span className="dietary">{label}</span></span>
            ))}
          </div>
        </div>
      </div>
    </div>
  );
};
