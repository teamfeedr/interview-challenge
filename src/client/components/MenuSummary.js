import React, { useEffect, useState } from "react";
import "../App.css";

export default ({ selectedList }) => {
  const [dietaries, setDietaries] = useState({});

  useEffect(() => {
    const listOfDietaries = selectedList.map((item) => item.dietaries);
    const dietariesObj = listOfDietaries.flat(1).reduce((acc, dietary) => {
      if (acc[dietary]) {
        acc[dietary] += 1;
      } else {
        acc[dietary] = 1;
      }
      return acc;
    }, {});
    setDietaries(dietariesObj);
  }, [selectedList]);

  return (
    <div className="menu-summary">
      <div className="container">
        <div className="row">
          <div className="col-6 menu-summary-left">
            <span>{selectedList.length} items</span>
          </div>
          <div className="col-6 menu-summary-right">
            {Object.keys(dietaries).map((dietary, index) => (
              <span key={index}>
                {dietaries[dietary]}x <span className="dietary">{dietary}</span>{" "}
              </span>
            ))}
          </div>
        </div>
      </div>
    </div>
  );
};
