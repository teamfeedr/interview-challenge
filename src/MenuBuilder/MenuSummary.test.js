import React from 'react';
import MenuSummary from './MenuSummary';

import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });

it('renders empty', () => {
  const div = shallow(<MenuSummary />);
  expect(div).toMatchSnapshot();
});

it('renders the total count of menu items and a list of dietery types along with their counts', () => {
  const dietaries = {
    v: 1,
    'n!': 2
  };
  const div = shallow(<MenuSummary count={1} dietaries={dietaries} />);
  expect(div).toMatchSnapshot();
});
