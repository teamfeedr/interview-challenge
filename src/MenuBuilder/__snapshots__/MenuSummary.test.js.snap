// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders empty 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MenuSummary />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        className="container"
>
        <div
                className="row"
        >
                <div
                        className="col-6 menu-summary-left"
                >
                        <span>
                                0
                                 items
                        </span>
                </div>
                <div
                        className="col-6 menu-summary-right"
                >
                        
                </div>
        </div>
</div>,
      "className": "menu-summary",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="row"
>
          <div
                    className="col-6 menu-summary-left"
          >
                    <span>
                              0
                               items
                    </span>
          </div>
          <div
                    className="col-6 menu-summary-right"
          >
                    
          </div>
</div>,
        "className": "container",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="col-6 menu-summary-left"
>
              <span>
                            0
                             items
              </span>
</div>,
            <div
              className="col-6 menu-summary-right"
>
              
</div>,
          ],
          "className": "row",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <span>
                0
                 items
</span>,
              "className": "col-6 menu-summary-left",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  0,
                  " items",
                ],
              },
              "ref": null,
              "rendered": Array [
                0,
                " items",
              ],
              "type": "span",
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [],
              "className": "col-6 menu-summary-right",
            },
            "ref": null,
            "rendered": Array [],
            "type": "div",
          },
        ],
        "type": "div",
      },
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="container"
>
          <div
                    className="row"
          >
                    <div
                              className="col-6 menu-summary-left"
                    >
                              <span>
                                        0
                                         items
                              </span>
                    </div>
                    <div
                              className="col-6 menu-summary-right"
                    >
                              
                    </div>
          </div>
</div>,
        "className": "menu-summary",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="row"
>
            <div
                        className="col-6 menu-summary-left"
            >
                        <span>
                                    0
                                     items
                        </span>
            </div>
            <div
                        className="col-6 menu-summary-right"
            >
                        
            </div>
</div>,
          "className": "container",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="col-6 menu-summary-left"
>
                <span>
                                0
                                 items
                </span>
</div>,
              <div
                className="col-6 menu-summary-right"
>
                
</div>,
            ],
            "className": "row",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <span>
                  0
                   items
</span>,
                "className": "col-6 menu-summary-left",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    0,
                    " items",
                  ],
                },
                "ref": null,
                "rendered": Array [
                  0,
                  " items",
                ],
                "type": "span",
              },
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [],
                "className": "col-6 menu-summary-right",
              },
              "ref": null,
              "rendered": Array [],
              "type": "div",
            },
          ],
          "type": "div",
        },
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`renders the total count of menu items and a list of dietery types along with their counts 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MenuSummary
    count={1}
    dietaries={
        Object {
            "n!": 2,
            "v": 1,
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        className="container"
>
        <div
                className="row"
        >
                <div
                        className="col-6 menu-summary-left"
                >
                        <span>
                                1
                                 items
                        </span>
                </div>
                <div
                        className="col-6 menu-summary-right"
                >
                        <span>
                                1
                                x 
                                <span
                                        className="dietary"
                                >
                                        v
                                </span>
                        </span>
                        <span>
                                2
                                x 
                                <span
                                        className="dietary"
                                >
                                        n!
                                </span>
                        </span>
                </div>
        </div>
</div>,
      "className": "menu-summary",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="row"
>
          <div
                    className="col-6 menu-summary-left"
          >
                    <span>
                              1
                               items
                    </span>
          </div>
          <div
                    className="col-6 menu-summary-right"
          >
                    <span>
                              1
                              x 
                              <span
                                        className="dietary"
                              >
                                        v
                              </span>
                    </span>
                    <span>
                              2
                              x 
                              <span
                                        className="dietary"
                              >
                                        n!
                              </span>
                    </span>
          </div>
</div>,
        "className": "container",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="col-6 menu-summary-left"
>
              <span>
                            1
                             items
              </span>
</div>,
            <div
              className="col-6 menu-summary-right"
>
              <span>
                            1
                            x 
                            <span
                                          className="dietary"
                            >
                                          v
                            </span>
              </span>
              <span>
                            2
                            x 
                            <span
                                          className="dietary"
                            >
                                          n!
                            </span>
              </span>
</div>,
          ],
          "className": "row",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <span>
                1
                 items
</span>,
              "className": "col-6 menu-summary-left",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  1,
                  " items",
                ],
              },
              "ref": null,
              "rendered": Array [
                1,
                " items",
              ],
              "type": "span",
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <span>
                  1
                  x 
                  <span
                                    className="dietary"
                  >
                                    v
                  </span>
</span>,
                <span>
                  2
                  x 
                  <span
                                    className="dietary"
                  >
                                    n!
                  </span>
</span>,
              ],
              "className": "col-6 menu-summary-right",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": "v",
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    1,
                    "x ",
                    <span
                      className="dietary"
>
                      v
</span>,
                  ],
                },
                "ref": null,
                "rendered": Array [
                  1,
                  "x ",
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": "v",
                      "className": "dietary",
                    },
                    "ref": null,
                    "rendered": "v",
                    "type": "span",
                  },
                ],
                "type": "span",
              },
              Object {
                "instance": null,
                "key": "n!",
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    2,
                    "x ",
                    <span
                      className="dietary"
>
                      n!
</span>,
                  ],
                },
                "ref": null,
                "rendered": Array [
                  2,
                  "x ",
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": "n!",
                      "className": "dietary",
                    },
                    "ref": null,
                    "rendered": "n!",
                    "type": "span",
                  },
                ],
                "type": "span",
              },
            ],
            "type": "div",
          },
        ],
        "type": "div",
      },
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="container"
>
          <div
                    className="row"
          >
                    <div
                              className="col-6 menu-summary-left"
                    >
                              <span>
                                        1
                                         items
                              </span>
                    </div>
                    <div
                              className="col-6 menu-summary-right"
                    >
                              <span>
                                        1
                                        x 
                                        <span
                                                  className="dietary"
                                        >
                                                  v
                                        </span>
                              </span>
                              <span>
                                        2
                                        x 
                                        <span
                                                  className="dietary"
                                        >
                                                  n!
                                        </span>
                              </span>
                    </div>
          </div>
</div>,
        "className": "menu-summary",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="row"
>
            <div
                        className="col-6 menu-summary-left"
            >
                        <span>
                                    1
                                     items
                        </span>
            </div>
            <div
                        className="col-6 menu-summary-right"
            >
                        <span>
                                    1
                                    x 
                                    <span
                                                className="dietary"
                                    >
                                                v
                                    </span>
                        </span>
                        <span>
                                    2
                                    x 
                                    <span
                                                className="dietary"
                                    >
                                                n!
                                    </span>
                        </span>
            </div>
</div>,
          "className": "container",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="col-6 menu-summary-left"
>
                <span>
                                1
                                 items
                </span>
</div>,
              <div
                className="col-6 menu-summary-right"
>
                <span>
                                1
                                x 
                                <span
                                                className="dietary"
                                >
                                                v
                                </span>
                </span>
                <span>
                                2
                                x 
                                <span
                                                className="dietary"
                                >
                                                n!
                                </span>
                </span>
</div>,
            ],
            "className": "row",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <span>
                  1
                   items
</span>,
                "className": "col-6 menu-summary-left",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    1,
                    " items",
                  ],
                },
                "ref": null,
                "rendered": Array [
                  1,
                  " items",
                ],
                "type": "span",
              },
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <span>
                    1
                    x 
                    <span
                                        className="dietary"
                    >
                                        v
                    </span>
</span>,
                  <span>
                    2
                    x 
                    <span
                                        className="dietary"
                    >
                                        n!
                    </span>
</span>,
                ],
                "className": "col-6 menu-summary-right",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": "v",
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      1,
                      "x ",
                      <span
                        className="dietary"
>
                        v
</span>,
                    ],
                  },
                  "ref": null,
                  "rendered": Array [
                    1,
                    "x ",
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "children": "v",
                        "className": "dietary",
                      },
                      "ref": null,
                      "rendered": "v",
                      "type": "span",
                    },
                  ],
                  "type": "span",
                },
                Object {
                  "instance": null,
                  "key": "n!",
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      2,
                      "x ",
                      <span
                        className="dietary"
>
                        n!
</span>,
                    ],
                  },
                  "ref": null,
                  "rendered": Array [
                    2,
                    "x ",
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "children": "n!",
                        "className": "dietary",
                      },
                      "ref": null,
                      "rendered": "n!",
                      "type": "span",
                    },
                  ],
                  "type": "span",
                },
              ],
              "type": "div",
            },
          ],
          "type": "div",
        },
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
